File: python/algos/path_sum.py
from __future__ import annotations


class TreeNode:
    def __init__(self, value: int = 0, left: TreeNode | None = None, right: TreeNode | None = None) -> None:
        self.val = value
        self.left = left
        self.right = right


def path_sum(node: TreeNode | None, target_sum: int) -> bool:
    if node is None:
        return False
    elif node.right is None and node.left is None:
        return node.val == target_sum
    else:
        return path_sum(node.left, target_sum - node.val) or path_sum(node.right, target_sum - node.val)

File: rust/src/algos/path_sum.rs
use std::cell::RefCell;
use std::cmp::min;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
    match root {
        Some(node) => match (node.borrow().left.clone(), node.borrow().right.clone()) {
            (None, None) => node.borrow().val == target_sum,
            _ => {
                path_sum(node.borrow().left.clone(), target_sum - node.borrow().val)
                    || path_sum(node.borrow().right.clone(), target_sum - node.borrow().val)
            }
        },
        None => false,
    }
}
